### ISSUE TITLE
LM-opencl crashes after a while

### ISSUE THREAD

**Issue by magnumripper** [magnumripper/JohnTheRipper#1790]
This on Titan X. Running LM-opencl against a couple hundred thousand hashes, cracking many of them, then:
```
Something went wrong while preparing hashes..Exiting..
*** Error in `../run/john': free(): corrupted unsorted chunks: 0x00000000044000f0 ***
Aborted (core dumped)
```
Resuming takes me a bit further each time.

**Comment by Sayantan2048** [comment]
Will investigate..

**Comment by Sayantan2048** [comment]
Can you give me the hash list and the options used ?  I suspect many more of the new formats could be affected!!

**Comment by magnumripper** [comment]
Sorry, no. Just build a test list out of rockyou.lst and pass_gen.pl or something!

**Comment by magnumripper** [comment]
The options were basically pure mask mode.

**Comment by Sayantan2048** [comment]
@magnumripper  Can we assume salt->count is updated during cracking and at the same time cracked binaries are removed from database every time we crack?

**Comment by magnumripper** [comment]
You should ask Solar on john-dev. I think it may be post-poned in some cases, at least in single mode. IIRC a binary pointed to by a salt hash can be set to NULL for some time before actually removed. But don't take my word on it.

**Comment by magnumripper** [comment]
> I suspect many more of the new formats could be affected!!

You really should move all duplicated code to either static functions in headers or actually shared functions. Like `select_bitmap()`, `prepare_table()` and `prepare_bitmap_*()`.

**Comment by Sayantan2048** [comment]
The problem is they are similar but not same. More work is required to unify them all.

**Comment by Sayantan2048** [comment]
fixed 569daf1c332c6317.

**Comment by magnumripper** [comment]
Good stuff!
